### DISCUSSION TITLE
How to generate random numbers in egui?

### DISCUSSION THREAD

**Discussion by gravit22** [emilk/egui#311] [Q&A] [answered]
I tried to use rand library, but it seems like egui doesn't support it. What can I use instead?

**Comment by lunabunn** [comment, marked as answer]
I'm curious, can you elaborate a bit on what you tried to reach the conclusion that egui doesn't support rand? I'm not an expert but I don't see how that would work.

**Reply by gravit22** [reply]
It fails when I am trying to use this function:
rand::thread_rng().gen_bool(0.2);
Try by yourself if you don't believe me.
I used it on egui_templete and last rand crate(0.8.3)

**Reply by lunabunn** [reply]
"It fails" is not very helpful, can't reproduce. What's the actual error you get?

**Reply by gravit22** [reply]
error: target is not supported, for more information see: https://docs.rs/getrandom/#unsupported-targets
   --> /home/mykhailo/.cargo/registry/src/github.com-1ecc6299db9ec823/getrandom-0.2.2/src/lib.rs:213:9
    |
213 | /         compile_error!("target is not supported, for more information see: \
214 | |                         https://docs.rs/getrandom/#unsupported-targets");
    | |_________________________________________________________________________^

error[E0433]: failed to resolve: use of undeclared crate or module `imp`
   --> /home/mykhailo/.cargo/registry/src/github.com-1ecc6299db9ec823/getrandom-0.2.2/src/lib.rs:235:5
    |
235 |     imp::getrandom_inner(dest)
    |     ^^^ use of undeclared crate or module `imp`

error: aborting due to 2 previous errors

For more information about this error, try `rustc --explain E0433`.
error: could not compile `getrandom`

To learn more, run the command again with --verbose.
warning: build failed, waiting for other jobs to finish...
error: build failed
![image](https://user-images.githubusercontent.com/33228017/115064762-fd5f7600-9ef5-11eb-93cd-2b7aad9dc625.png)

My actions:
Add rand crate to cargo.toml
add line I mentioned above.
./setup_web.sh
./start_server.sh
./build_web.sh

**Reply by lunabunn** [reply]
that doesnt look egui related. does a non egui project w/ rand build fine for the same target? 👀

**Reply by gravit22** [reply]
No. Only Egui project. When I compile it locally it works fine.
I don't use wasm for any other projects.

**Reply by lunabunn** [reply]
Right. Quick skim through the doc page I linked reveals wasm32-unknown-unknown w/ JS needs the `js` feature to be set. I imagine you didn't set this correctly?

**Reply by gravit22** [reply]
I have no idea.

**Reply by lunabunn** [reply]
Either way, AFAICT this has nothing to do with egui. It's just that getrandom doesn't support wasm32-unknown-unknown out of the box.

**Reply by gravit22** [reply]
When I added this line to dependencies, everything works.
getrandom = { version = "0.2", features = ["js"] }
